File: fabric/src/main/java/me/pandamods/pandalib/fabric/platform/RegistrationHelperImpl.java
/*
 * Copyright (C) 2024 Oliver Froberg (The Panda Oliver)
 *
 * This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  any later version.
 *
 * You should have received a copy of the GNU Lesser General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package me.pandamods.pandalib.fabric.platform;

import me.pandamods.pandalib.platform.services.RegistrationHelper;
import me.pandamods.pandalib.registry.DeferredObject;
import net.fabricmc.fabric.api.resource.IdentifiableResourceReloadListener;
import net.fabricmc.fabric.api.resource.ResourceManagerHelper;
import net.minecraft.core.RegistrationInfo;
import net.minecraft.core.Registry;
import net.minecraft.core.WritableRegistry;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.packs.PackType;
import net.minecraft.server.packs.resources.PreparableReloadListener;
import net.minecraft.server.packs.resources.ResourceManager;

import java.util.Collection;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;
import java.util.function.Supplier;

public class RegistrationHelperImpl implements RegistrationHelper {
	@Override
	@SuppressWarnings("unchecked")
	public <T> void register(DeferredObject<? extends T> deferredObject, Supplier<? extends T> supplier) {
		Registry.register((Registry<T>) deferredObject.getRegistry(), deferredObject.getId(), supplier.get());
	}

	@Override
	@SuppressWarnings({"unchecked", "rawtypes"})
	public <T> void registerNewRegistry(Registry<T> registry) {
		ResourceLocation registryName = registry.key().location();
		if (BuiltInRegistries.REGISTRY.containsKey(registryName))
			throw new IllegalStateException("Attempted duplicate registration of registry " + registryName);
		
		((WritableRegistry) BuiltInRegistries.REGISTRY).register(registry.key(), registry, RegistrationInfo.BUILT_IN);
	}
	
	@Override
	public void registerReloadListener(PackType packType, PreparableReloadListener listener, ResourceLocation id, List<ResourceLocation> dependencies) {
		ResourceManagerHelper.get(packType).registerReloadListener(new IdentifiableResourceReloadListener() {
			@Override
			public ResourceLocation getFabricId() {
				return id;
			}

			@Override
			public CompletableFuture<Void> reload(PreparationBarrier preparationBarrier, ResourceManager resourceManager, Executor executor, Executor executor2) {
				return listener.reload(preparationBarrier, resourceManager, executor, executor2);
			}

			@Override
			public Collection<ResourceLocation> getFabricDependencies() {
				return dependencies;
			}

			@Override
			public String getName() {
				return listener.getName();
			}
		});
	}
}

File: fabric/src/main/kotlin/dev/pandasystems/pandalib/fabric/platform/RegistrationHelperImpl.kt
/*
 * Copyright (C) 2024 Oliver Froberg (The Panda Oliver)
 *
 * This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  any later version.
 *
 * You should have received a copy of the GNU Lesser General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package dev.pandasystems.pandalib.fabric.platform

import dev.pandasystems.pandalib.platform.services.RegistrationHelper
import dev.pandasystems.pandalib.registry.DeferredObject
import net.fabricmc.fabric.api.resource.IdentifiableResourceReloadListener
import net.fabricmc.fabric.api.resource.ResourceManagerHelper
import net.minecraft.core.RegistrationInfo
import net.minecraft.core.Registry
import net.minecraft.core.WritableRegistry
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.resources.ResourceKey
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.packs.PackType
import net.minecraft.server.packs.resources.PreparableReloadListener
import net.minecraft.server.packs.resources.ResourceManager
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executor
import java.util.function.Supplier

class RegistrationHelperImpl : RegistrationHelper {
	override fun <T> register(deferredObject: DeferredObject<out T>, supplier: Supplier<out T>) {
		@Suppress("UNCHECKED_CAST")
		Registry.register(deferredObject.registry as Registry<T>, deferredObject.id, supplier.get())
	}

	override fun <T> registerNewRegistry(registry: Registry<T>) {
		val registryName = registry.key().location()
		check(BuiltInRegistries.REGISTRY.containsKey(registryName)) { "Attempted duplicate registration of registry $registryName" }

		@Suppress("UNCHECKED_CAST")
		(BuiltInRegistries.REGISTRY as WritableRegistry<Registry<*>>).register(registry.key() as ResourceKey<Registry<*>>, registry, RegistrationInfo.BUILT_IN)
	}

	override fun registerReloadListener(
		packType: PackType,
		listener: PreparableReloadListener,
		id: ResourceLocation,
		dependencies: Collection<ResourceLocation>
	) {
		ResourceManagerHelper.get(packType).registerReloadListener(object : IdentifiableResourceReloadListener {
			override fun getFabricId(): ResourceLocation {
				return id
			}

			override fun reload(
				preparationBarrier: PreparableReloadListener.PreparationBarrier,
				resourceManager: ResourceManager,
				backgroundExecutor: Executor,
				gameExecutor: Executor
			): CompletableFuture<Void> {
				return listener.reload(preparationBarrier, resourceManager, backgroundExecutor, gameExecutor)
			}

			override fun getFabricDependencies(): Collection<ResourceLocation> {
				return dependencies
			}

			override fun getName(): String {
				return listener.name
			}
		})
	}
}
